### PR TITLE
IMU driver: follow ROS msg specs, set covariances

### PR DESCRIPTION
**Pull request by adivardi**
## Additions & Changes
Several modifications to the Livox driver related to the internal IMU driver and data publishing.
These changes are required for the robot_localization's EKF to use the IMU data.

Note: The gyroscope bias compensation and the external frame transformation were moved to the main repo. See PR https://github.com/enwaytech/cleansquare_ros/pull/4267.

- set orientation covariance following the ROS msg's specs to show no orientation data is available (1st element of covariance vector should be -1)
- Convert the linear acceleration units from g's to m/s2, following the ROS msg specs
- add covariances from config for linear acceleration and angular velocity
- Change the IMU frame to take into account the distance between the LIDAR center and the IMU center (translation taken from Livox user manual)
- make sure driver fails if the user_config parsing failed

Reverted:
- ~~Add an option to transform the IMU data to an external frame before publishing. This is needed as the EKF expects the IMU data to be in a frame with the same orientation as base_link.~~
- ~~Add an option to compensate gyroscope bias using static bias values. This is required since the Livox IMUs show high values of bias, especially for the angular velocity around x (in the order of 2-3 degrees/sec)~~

Test together with the main repo PR https://github.com/enwaytech/cleansquare_ros/pull/4267

Run on the robot and check that all the modifications took place (covariances, acceleration units, ~~frame transform, gyro bias,~~ ...)
You can plot the angular velocities in rqt_plot and make sure they are close to 0 over time (ideally, they should be in the order of 0.00X)

## Recommended reviewer
Georg?

### Checklist (tick off when done or remove when not relevant):
- [x] Set labels (do not forget a review time estimation label)
- [x] Set good PR title for changelog
- [x] Add package dependencies
- [x] Check the linter warnings in the PR checks or [locally](https://enwaytech.atlassian.net/wiki/spaces/ENWAY/pages/5308462/C+Coding+Guideline#Code-Checking:-Local-Lint-Reports)
- [x] Perform a self-review if the PR has more than 50 lines of changes

[sc-14152]
[sc-14159]
[sc-14176]

### CODE REVIEW

File: src/lddc.cpp [hunk 1]
```diff
@@ -46,16 +48,20 @@ namespace livox_ros
 
 /** Lidar Data Distribute Control--------------------------------------------*/
 #ifdef BUILDING_ROS1
-Lddc::Lddc(int format, int multi_topic, int data_src, int output_type,
-    double frq, std::string &frame_id, bool lidar_bag, bool imu_bag, bool dust_filter)
+Lddc::Lddc(int format, int multi_topic, int data_src, int output_type, double frq, std::string &frame_id,
+           const std::vector<double>& angular_velocity_covariance, const std::vector<double>& linear_acceleration_covariance,
+    bool lidar_bag, bool imu_bag, bool dust_filter)
```
Comment (adivardi) [inline src/lddc.cpp]: fix alignment

File: src/lddc.h [hunk 2]
```diff
@@ -144,6 +150,10 @@ class Lddc final {
   double publish_frq_;
   uint32_t publish_period_ns_;
   std::string frame_id_;
+  bool transform_imu_to_external_frame_;
```
Comment (adivardi) [inline src/lddc.h]: useless

File: src/lddc.h [hunk 3]
```diff
@@ -144,6 +150,10 @@ class Lddc final {
...
+  std::string external_frame_id_;
```
Comment (adivardi) [inline src/lddc.h]: also useless

File: src/lddc.cpp [hunk 4]
```diff
@@ -597,15 +600,31 @@ void Lddc::InitImuMsg(const ImuData& imu_data, ImuMsg& imu_msg, uint64_t& timest
...
+  imu_msg.linear_acceleration.x = imu_data.acc_x * g_to_ms2;
+  imu_msg.linear_acceleration.y = imu_data.acc_y * g_to_ms2;
+  imu_msg.linear_acceleration.z = imu_data.acc_z * g_to_ms2;
+
+  // set covariances from config for angular_velocity and linear_acceleration, and reset orientation_covariance
+  for(int i = 0; i < 9; i++)
+  {
+	  imu_msg.angular_velocity_covariance[i] = angular_velocity_covariance_[i];
+	  imu_msg.linear_acceleration_covariance[i] = linear_acceleration_covariance_[i];
+	  imu_msg.orientation_covariance[i] = 0;
```
Comment (adivardi) [inline src/lddc.cpp]: indentation is using tabs...

**Review by adivardi — COMMENTED** [review]
ok

**Comment by adivardi** [comment]
@georgflick I rebased and checked the changes are the same as before. Ready now

**Review by georgflick — APPROVED** [review]
Tested on v20e and worked fine.

**Pull request merged**